### DISCUSSION TITLE
Help to invert the uart signal in Arduino

### DISCUSSION THREAD

**Discussion by dzungpv** [revk/xESP32-Faikout#94] [Q&A] [answered]
Hello, I try to build Arduino firmware version, origin version work fine but I want to add some modification, the original source code too complex to understand with full of macro ... that why I rewire it in Arduino.
In the sketch I make it work with non invert test board (non invert tx rx), but not with origin board design with TX and RX invert with FETS.
I use HardwareSerial from official repo https://github.com/espressif/arduino-esp32/blob/master/cores/esp32/HardwareSerial.cpp
and connect to S21 with:

```
_HardSerial->begin(2400, SERIAL_8E2, rx, tx, true); //Serial 1, rx: 27, tx: 26
```
But it never work, I also try to use -27 and -26.

**Comment by revk** [comment]
In my code the negative port numbers are not actually negative numbers. The - is just a flag in the settings to tell my code it is inverted. You set the port (positive number) and set the serial to be inverted using a separate call. Arduino must have a call to set inverted serial signals.

**Reply by dzungpv** [reply]
In the code:
```
 uart_set_pin (uart, port_mask (tx), port_mask (rx), -1, -1);
```
 and
```
 #define port_mask(p) ((p)&0x3F)
```
 you just convert the negative port to normal port?
And what is the code:  rx & PORT_INV) (#define PORT_INV 0x40) do? I check the hal code of the ESP UART but not see code like that, so that why it is not working.
They only write code like this:
```
    // Is it right or the idea is to swap rx and tx pins? 
    if (inverted) {
        // invert signal for both Rx and Tx
        ESP_ERROR_CHECK(uart_set_line_inverse(uart_nr, UART_SIGNAL_TXD_INV | UART_SIGNAL_RXD_INV));    
    }
```
So you are trying to invert the rx/tx independent?

**Reply by revk** [reply]
My code sets 0x40 in the port number if prefixed with a -.

Yes `uart_set_line_inverse` looks like it is what you use to invert.

**Reply by dzungpv** [reply]
Yes, but what about your  port_mask (tx) code does?

**Comment by revk** [comment, marked as answer]
port_mask just gets the actual GPIO number to use, tripping the extra bits I set for inverted, etc. If you are not using my code and just have port numbers, use them!

**Reply by dzungpv** [reply]
Yes, I am not use it directly but I will adapt it in Arduino.
